src/concurrency/timestamp_ordering_transaction_manager.cpp

```diff
@@ -778,130 +778,136 @@ ResultType TimestampOrderingTransactionManager::CommitTransaction(
   oid_t database_id = 0;
   if (static_cast<StatsType>(settings::SettingsManager::GetInt(settings::SettingId::stats_mode)) !=
       StatsType::INVALID) {
-    if (!rw_set.empty()) {
-      database_id =
-          manager.GetTileGroup(rw_set.begin()->first)->GetDatabaseId();
+    if (!rw_set.IsEmpty()) {
+      // Call the GetConstIterator() function to explicitly lock the cuckoohash
+      // and initilaize the iterator
+      auto rw_set_lt = rw_set.GetConstIterator();
+      const auto tile_group_id = rw_set_lt.begin()->first.block;
+      database_id = manager.GetTileGroup(tile_group_id)->GetDatabaseId();
     }
   }
 
   // install everything.
   // 1. install a new version for update operations;
   // 2. install an empty version for delete operations;
   // 3. install a new tuple for insert operations.
-  for (auto &tile_group_entry : rw_set) {
-    oid_t tile_group_id = tile_group_entry.first;
-    auto tile_group_header = manager.GetTileGroup(tile_group_id)->GetHeader();
+  // Iterate through each item pointer in the read write set
+
+  // TODO (Pooja): This might be inefficient since we will have to get the
+  // tile_group_header for each entry. Check if this needs to be consolidated
+  for (const auto &tuple_entry : rw_set.GetConstIterator()) {
+    ItemPointer item_ptr = tuple_entry.first;
+    oid_t tile_group_id = item_ptr.block;
+    oid_t tuple_slot = item_ptr.offset;
 
-    for (auto &tuple_entry : tile_group_entry.second) {
-      auto tuple_slot = tuple_entry.first;
+    auto tile_group_header = manager.GetTileGroup(tile_group_id)->GetHeader();
```

Comment:
I suppose `manager.GetTileGroup()` needs to acquire read lock on the catalog? If yes, then better move this line of code out of the for loop. I wrote the code like this as at that time we only support static catalog table.

You could also manage the rw_set as `map<table_id, map<tuple_id, tuple>>`. In this case, you could easily cache the tile group header for each table.

Reply:
Actually the catalog uses a lock_free_array to store the tile_group_id -> tile_group mapping. Once we move to using tbb::concurrent_vector, that should not be a problem because the `at()` function does not acquire a lock.

Reply:
In fact, most of the concurrent data structures still require use atomic operations to process read operations. In `concurrent_vector`, maybe it maintains a variable `offset` to keep track of the boundary for the visible slot ranges. You may do some simple experiments to evaluate whether read is scalable or not. But AFAIK, the scalability of the concurrent vector is unlikely to be a problem in Peloton as some other components (e.g., frontend) can bring much more overhead :-)